### ISSUE TITLE
Pagination of primary data not supported

### ISSUE THREAD

**Issue by emetsger** [jasminb/jsonapi-converter#33]
It seems pagination of primary data is not supported; this is because `readObjectCollection` returns `List<T>` - the return type is not able to expose pagination links (or `meta` information, for that matter).

The JSON below represents paginated primary data.  The total size of the primary data is `m` but only `n` results are returned:

    {
        "data": [
            {
              // object 1
            },
            { 
              // object 2
            },
            {
              // object n
            }
        ],
        "links": {
            "first": null,
            "last": "http://example.com/nodes/?page=m",
            "prev": null,
            "next": "http://example.com/nodes/?page=2",
        }
    }

The `ResourceConverter` will represent the first page of results in a `List<T>` with a `size()` of `n`.  The client has no way of knowing if the results are paginated or not, because  `List` doesn't expose any pagination state.

Per the [spec](http://jsonapi.org/format/#fetching-pagination), pagination **MAY** be supported, and if so, **MUST** be represented by a top-level `links` object:

    Pagination
    
    A server MAY choose to limit the number of resources returned in a response to a subset (“page”) of the whole set available.
    
    A server MAY provide links to traverse a paginated data set (“pagination links”).
    
    Pagination links MUST appear in the links object that corresponds to a collection. To paginate the primary data, supply pagination links in the top-level links object. To paginate an included collection returned in a compound document, supply pagination links in the corresponding links object.
    
    The following keys MUST be used for pagination links:
    
      * first: the first page of data
      * last: the last page of data
      * prev: the previous page of data
      * next: the next page of data
    
    Keys MUST either be omitted or have a null value to indicate that a particular link is unavailable.

**Comment by emetsger** [comment]
Here are a couple of proposals to handle this.  What do you think?

### Supply a `List` factory to `ResourceConverter`

The `ResourceConverter` uses a factory interface for producing the `List` instances that are returned by `readObjectCollection`.  The factory interface provides access to the top-level `links` and `meta` so the caller can return a custom `List` implementation.  If the caller doesn't supply an implementation, the default implementation is used by `ResourceConverter` and returns an `ArrayList`.

### Update `ResourceConverter` to expose pagination state

Because pagination is a part of the JSON API specification, it would be reasonable to have `readObjectCollection` expose pagination state.  `readObjectCollection` could be updated to return a `PaginatedList`:

    <T> PaginatedList<T> readObjectCollection(...)

```java
    public class PaginatedList<E> extends AbstractList {
    
        private String prev;
    
        private String next;
    
        private String first;
    
        private String last;
    
        private List<E> objects;
    
        public String getPrev() {
            return prev;
        }
    
        public void setPrev(String prev) {
            this.prev = prev;
        }
    
        public String getNext() {
            return next;
        }
    
        public void setNext(String next) {
            this.next = next;
        }
    
        public String getFirst() {
            return first;
        }
    
        public void setFirst(String first) {
            this.first = first;
        }
    
        public String getLast() {
            return last;
        }
    
        public void setLast(String last) {
            this.last = last;
        }
    
        public List<E> getObjects() {
            return objects;
        }
    
        public void setObjects(List<E> objects) {
            this.objects = objects;
        }
    
        @Override
        public E get(int index) {
            return objects.get(index);
        }
    
        @Override
        public int size() {
            return objects.size();
        }
    
    }
```

**Comment by jasminb** [comment]
Hey, similar problem is also present with meta, where it is not possible to expose it without having some kind of collection wrapper.

So, instead of having `PaginatedList`, something like `ResourceCollection` class could be implemented that would allow exposing both pagination and meta objects.

**Comment by emetsger** [comment]
Ok, thanks, I'll take a look at it!

**Comment by justinharrison** [comment]
Any updates on pagination support? Currently this is the most complete implementation of jsonapi.org's spec on Java and I'd like to use it but I can't without pagination support.

**Comment by jasminb** [comment]
Should be on develop in next few days.

**Comment by emetsger** [comment]
@jasminb

Any thoughts on my `ResourceList` implementation?  It may not be the best
name.  I avoided the name `ResourceCollection` because I didn't want to
conflate the Java concept of Collection with the JSON-API concept of
collection (and besides, the method returns a `List` anyway).

**Comment by bradtofel** [comment]
I think the PaginatedList(/ResourceList) solution will cover some cases, but I'm not sure if it handles everything.

According to the spec, it's possible to have a 'links' field for each Resource.

So we could have:

A:
{
  'links' :{},
  'data' :{}
}

B:
{
  'data' :{'links':{}}
}

C:
{
  'data' :[{'links':{}}]
}


In the above scenario, you can also replace 'links' with 'meta'. (Maybe we leave the 'meta' on 'links' out of scope for now?)

Currently in master, (I think) only 'A' is handled for Resource request, by placing the 'meta' data onto the returned Resource (via annotation.) This has major simplicity advantages. PaginatedList would also handle 'A', but I don't think scenarios 'B' or 'C' are handled.

Using @Links and @Meta annotations on Resource POJOs seems very natural and simple, and to me feels like a very good fit for scenarios 'B' and 'C', where the 'links' (or 'meta') data applies to the Resource.

I'm wondering if it makes sense to add a top-level response object, call it maybe JSONAPIResponse, or JSONAPIDocument, which is always the target container for meta/links at the top level (scenario 'A') as well as for the 'data' field, for both Resource and collection responses. It would also be a natural place to put 'errors', 'jsonapi' down the road.


It seems like we could support current interface to ResourceConverter, for those who don't need the extra layer:

```java
    public <T> JsonApiDocument<List<T>> readCollectionDocument(byte[] bytes, Class<T> clazz) {
	// magic happens...
    }

    public <T> List<T> readObjectCollection(byte[] bytes, Class<T> clazz) {

        return readCollectionDocument(bytes, clazz).getData();
    }
```

**Comment by emetsger** [comment]
@bradtofel I agree that a higher-level abstraction is needed, and I like the idea of the `JsonApiDocument`.  It would maintain binary compatibility and offer a path forward for the optional members in the response (like `jsonapi`, etc.).

**Comment by jasminb** [comment]
@emetsger will take the code you written to handle links and incorporate it into the proposal from @bradtofel.

**Comment by bradtofel** [comment]
hey @jasminb was just getting that PR together. : )

Feel free to grab any bits from there that look like they'd fit as well.

**Comment by jasminb** [comment]
@bradtofel will to, thanks!

**Comment by jasminb** [comment]
@emetsger @bradtofel Please see https://github.com/jasminb/jsonapi-converter/pull/41

This is just a placeholder PR for ongoing work that will be performed on the given branch.

Currently working (not unit-tested):

- top level and resource level meta
- top level links

Resource level links is still work in progress.

**Comment by bradtofel** [comment]
@jasminb I'm not sure I follow what you mean about static? Do you mean to deal with the 'links' from scenario 'B' or 'C', you'd have to have a member of a type from this library to deal with it, like below?

```
class Bar {
  @Links
  com.github.jasminb.jsonapi.LinksData barLinks;
  ...
}
```

**Comment by emetsger** [comment]
@jasminb Yes, I'm with @bradtofel regarding `@Links`:

```json
{
    "data": [
        {            
            "links": {
                "info": "http://localhost:8000/v2/files/57347c2ecfa27c00468cedc3/",
                "self": "http://localhost:8000/v2/files/57347c2ecfa27c00468cedc3/",
                "move": "http://localhost:7777/v1/resources/pd24n/providers/osfstorage/57347c2ecfa27c00468cedc3",
                "upload": "http://localhost:7777/v1/resources/pd24n/providers/osfstorage/57347c2ecfa27c00468cedc3",
                "download": "http://localhost:7777/v1/resources/pd24n/providers/osfstorage/57347c2ecfa27c00468cedc3",
                "delete": "http://localhost:7777/v1/resources/pd24n/providers/osfstorage/57347c2ecfa27c00468cedc3"
            },
            "attributes": {
                "extra": {
                    "hashes": {
                        "sha256": "39e810498a96517091cac7f59abb3a57d115f3c8b8b1e22648426dfe66bf9cd7",
                        "md5": "19d61973a43bf1550aef3241b5543076"
                    }
                },
                "kind": "file",
                "name": "porsche.jpg",
                "last_touched": null,
                "materialized_path": "/porsche.jpg",
                "date_modified": "2016-05-12T12:50:54.488Z",
                "date_created": "2016-05-12T12:50:54.488Z",
                "provider": "osfstorage",
                "path": "/57347c2ecfa27c00468cedc3",
                "checkout": null,
                "size": 16064286
            },
            "type": "files",
            "id": "57347c2ecfa27c00468cedc3"
        }
    ],
    "links": {
        "first": null,
        "last": null,
        "prev": null,
        "next": null
    }
}
```

The resource in the `data` member above models a domain entity `File`.  The `data.links` member contains URLs used to manipulate the file.  So in my Java class for the domain entity, I might have a member like: 

```java
@Links
Map<String, ?>
```

Essentially the same as `@Meta`.  As an API designer, I may use `@Links` just as I would use `@Meta`.  The reasoning used to support an `@Meta` annotation is the same reasoning you would use to argue for an `@Links` annotation.

**Comment by emetsger** [comment]
See [this](https://github.com/emetsger/jsonapi-converter/blob/79d00e4bc80c4f64e3a1ca1a817755bf287e341d/src/test/java/com/github/jasminb/jsonapi/models/link/NodeLinksAsDomainObject.java) in my [link-annotation](https://github.com/emetsger/jsonapi-converter/tree/link-annotation) branch: `@Links` can be used to map to a domain object or to a `Map<String, ?>`.

**Comment by jasminb** [comment]
Hey @emetsger, I've merged #41. Please check it out and close the issue.
